### ISSUE TITLE
`TypeVar`s from different scopes don't unify

### ISSUE THREAD

**Issue by colehaus** [microsoft/pyright#3851]
All of the code below type checks fine in mypy. In pyright 1.1.267, `b` type checks but `x` does not.
```py
from typing import Protocol, TypeVar

X = TypeVar("X")

class Transform(Protocol):
    def __call__(self, x: X) -> X: ...

def b(a: Transform, y: X) -> X:
    return a(y)

def x(y: X) -> X:
    t: Transform = lambda x: x
    return t(y)
```

I would expect `x` to type check, but instead the final line produces the error:

```
13:14 - error: Argument of type "X@x" cannot be assigned to parameter "x" of type "X@__call__"
    Type "X@x" cannot be assigned to type "X@__call__" (reportGeneralTypeIssues)
13:12 - error: Expression of type "X@__call__" cannot be assigned to return type "X@x"
    Type "X@__call__" cannot be assigned to type "X@x" (reportGeneralTypeIssues)
```

**Comment by erictraut** [comment]
Pyright's behavior here is by design.

This is discussed in detail in this issue: https://github.com/microsoft/pyright/issues/3788. The more relevant part is [this post](https://github.com/microsoft/pyright/issues/3788#issuecomment-1206802620).

**Comment by colehaus** [comment]
Oh, interesting. Thanks for the response!

It somehow have the sense that this idea should be expressible. Here's the analogous Haskell for example:
```hs
f :: a -> a
f a = g a
  where
    g :: x -> x
    g x = x
```
I suppose the closest thing that's currently supported in pyright is:
```py
from typing import Protocol, TypeVar

X = TypeVar("X")

class Transform(Protocol[X]):
    def __call__(self, x: X) -> X: ...

def b(a: Transform[X], y: X) -> X:
    return a(y)

def x(y: X) -> X:
    t: Transform[X] = lambda x: x
    return t(y)
```
?

**Comment by erictraut** [comment]
This is an aspect of the Python static type system that is not well defined or specified. PEP 484 introduced some rules around [TypeVar scoping](https://peps.python.org/pep-0484/#scoping-rules-for-type-variables), but it didn't anticipate or discuss some of the edge cases. PEP 544 introduced the notion of a [callback protocol](https://peps.python.org/pep-0544/#callback-protocols), but it didn't discuss generics in callback protocols.

If you feel strongly about this particular use case, you could champion a new extension to the Python type system (probably in the form of a new PEP) that clarifies and standardizes the behavior you're looking for. The best forum to discuss such proposals is either the [python/typing discussion forum](https://github.com/python/typing/discussions) or the [typing-sig email group](https://mail.python.org/archives/list/typing-sig@python.org/).
